File: app/src/main/java/org/test/koltintest/UsersFragment.java
package org.test.koltintest;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import org.test.koltintest.model.User;

import kotlin.Unit;

public class UsersFragment extends Fragment {

    public static UsersFragment create() {
        return new UsersFragment();
    }

    {
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_users, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        view.<RecyclerView>findViewById(R.id.recycler).setAdapter(new UsersAdapter(this::onUserSelected));
    }

    private void onUserSelected(User user) {
        getFragmentManager().beginTransaction() // TODO podría hacerse genérico
                .replace(getId(), UsersEditFragment.create(user))
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        inflater.inflate(R.menu.fragment_users, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add:
                getFragmentManager().beginTransaction() // TODO podría hacerse genérico
                        .replace(getId(), UsersEditFragment.create())
                        .addToBackStack(null)
                        .commit();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/org/test/koltintest/UsersFragment.kt
package org.test.koltintest

import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_users.*
import org.test.koltintest.model.User
import org.test.replaceWith

class UsersFragment : Fragment() {

    init {
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_users, container, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recycler.adapter = UsersAdapter(this::onUserSelected)
    }

    private fun onUserSelected(user: User) {
        replaceWith(UsersEditFragment.create(user))
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)

        inflater.inflate(R.menu.fragment_users, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add -> {
                replaceWith(UsersEditFragment.create())
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {

        fun create() = UsersFragment()

    }

}
